### PR TITLE
CXX-530 remove string_or_literal

### PR DESCRIPTION
**Pull request by amidvidy**

### CODE REVIEW

File: src/bsoncxx/builder/core.hpp [hunk 1]
```diff
@@ -23,6 +23,7 @@
 #include <bsoncxx/array/view.hpp>
 #include <bsoncxx/document/value.hpp>
 #include <bsoncxx/document/view.hpp>
+#include <bsoncxx/stdx/string_view.hpp>
```
Comment (TylerBrock) [inline src/bsoncxx/builder/core.hpp]: Could possibly forward declare here?

File: src/bsoncxx/document/element.cpp [hunk 2]
```diff
@@ -65,7 +65,7 @@ string_or_literal element::key() const {
 
     const char* key = bson_iter_key(&iter);
 
-    return string_or_literal{key, std::strlen(key)};
+    return stdx::string_view{key, std::strlen(key)};
```
Comment (acmorrow) [inline src/bsoncxx/document/element.cpp]: Do you need the strlen here?

File: src/bsoncxx/document/element.cpp [hunk 3]
```diff
@@ -144,8 +144,8 @@ types::b_regex element::get_regex() const {
     const char* options;
     const char* regex = bson_iter_regex(&iter, &options);
 
-    return types::b_regex{string_or_literal{regex, std::strlen(regex)},
-                          string_or_literal{options, std::strlen(options)}};
+    return types::b_regex{stdx::string_view{regex, std::strlen(regex)},
```
Comment (acmorrow) [inline src/bsoncxx/document/element.cpp]: Ditto re strlen, everywhere

File: src/bsoncxx/builder/core.hpp (hunk 1 above)
Comment (amidvidy) [inline src/bsoncxx/builder/core.hpp]: actually this is completely unnecessary

**Comment by amidvidy** [comment]
@hanumantmk, how does this look

File: src/bsoncxx/test/bson_builder.cpp [hunk 4]
```diff
@@ -583,10 +583,29 @@ TEST_CASE("array builder finalizes", "[bsoncxx::builder::stream]") {
...
 
+TEST_CASE("document core builder ownership", "[bsoncxx::builder::core]") {
+    builder::core b(false);
+
+    SECTION("when passing a std::string key, ownership transfers to the instance") {
+        {
+            std::string key{"falafel"};
+            b.key_owned(key);
+        }
+        b.append(1);
```
Comment (amidvidy) [inline src/bsoncxx/test/bson_builder.cpp]: TODO, actually assert that the builder's view contains the correct keys.

File: src/bsoncxx/builder/core.cpp [hunk 5]
```diff
@@ -178,7 +185,9 @@ class core::impl {
     bson_t _root;
 
     itoa _itoa_key;
-    string_or_literal _user_key;
+
+    stdx::string_view _user_key_view;
+    std::string _user_key_owning;
```
Comment (amidvidy) [inline src/bsoncxx/builder/core.cpp]: note to self, rename to _user_key_owned for consistency

File: src/bsoncxx/builder/stream/key_context.hpp [hunk 6]
```diff
@@ -36,12 +37,12 @@ class key_context {
 
     template <std::size_t n>
     value_context<key_context> operator<<(const char (&v)[n]) {
-        _core->key_literal(v, n - 1);
+        _core->key_view(stdx::string_view{v, n - 1});
```
Comment (hanumantmk) [inline src/bsoncxx/builder/stream/key_context.hpp]: I suspect we'll want a stdx::string_view overload as well as the literal and std::string options

**Comment by hanumantmk** [comment]
My one critique would be that we probably want stdx::string_view overloads everywhere we have std::string overloads for the basic and stream builder apis.  As it is, we'll only pick up literals and std::strings, so we're probably doing some unnecessary copying if the user actually has string_views from some other source.

Other than that, LGTM

**Pull request merged**